k2/csrc/pytorch_context.h

```diff
@@ -15,10 +13,85 @@
 #ifndef K2_CSRC_PYTORCH_CONTEXT_H_
 #define K2_CSRC_PYTORCH_CONTEXT_H_
 
+#include "c10/cuda/CUDACachingAllocator.h"
 #include "k2/csrc/context.h"
+#include "k2/csrc/log.h"
+#include "torch/torch.h"
+
+namespace k2 {
 
 class PytorchContext : public Context {
-  // ...
+ public:
+  // if device_id < 0, then this is a cpu context;
+  // otherwise, it is a cuda context.
+  explicit PytorchContext(int32_t device_id) : device_id_(device_id) {
+    if (device_id_ < 0)
+      InitCpu();
+    else
+      InitCuda();
+  }
+
+  ContextPtr GetCpuContext() override {
+    // TODO(fangjun): return `this` if it's cpu ?
+    return nullptr;
+  }
+
+  ContextPtr GetPinnedContext() override { return nullptr; }
+
+  DeviceType GetDeviceType() const override {
+    return device_id_ >= 0 ? kCuda : kCpu;
+  }
+
+  int32_t GetDeviceId() const override { return device_id_; }
+
+  cudaStream_t GetCudaStream() const override {
+    return device_id_ >= 0 ? c10::cuda::getCurrentCUDAStream(device_id_)
+                           : kCudaStreamInvalid;
+  }
+
+  void *Allocate(std::size_t bytes, void **deleter_context) override {
+    void *p = allocator_->raw_allocate(bytes);
+    if (deleter_context) *deleter_context = nullptr;
+    return p;
+  }
+
+  void Deallocate(void *data, void * /*deleter_context*/) override {
+    allocator_->raw_deallocate(data);
+  }
+
+  bool IsCompatible(const Context &other) const override {
+    return other.GetDeviceType() == GetDeviceType() &&
+           other.GetDeviceId() == device_id_;
+  }
+
+  void Sync() const override {
+    if (device_id_ >= 0) {
+      auto ret = cudaStreamSynchronize(GetCudaStream());
+      K2_CHECK_CUDA_ERROR(ret);
+    }
+  }
+
+ private:
+  void InitCpu() {
+    allocator_ = torch::GetAllocator(torch::kCPU);
+    K2_CHECK(allocator_->raw_deleter() != nullptr);
+  }
+
+  void InitCuda() {
+    auto ret = cudaSetDevice(device_id_);
+    K2_CHECK_CUDA_ERROR(ret);
+    // TODO(fangjun): invoke init only once
+    c10::cuda::CUDACachingAllocator::init(device_id_ + 1);
```

Comment:
we might have to introduce some static variables to avoid multiple calls to init..

Reply:
PyTorch uses `std::call_once` to achieve this.

Relevant codes are:

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Context.h#L78
```cpp
  THCState* lazyInitCUDA() {
    std::call_once(thc_init,[&] {
      thc_state = detail::getCUDAHooks().initCUDA();
    });
    return thc_state.get();
  }
```

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/detail/CUDAHooks.cpp#L45
```cpp
std::unique_ptr<THCState, void (*)(THCState*)> CUDAHooks::initCUDA() const {
  THCudaInit(thc_state);
}
```

https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCGeneral.cpp#L41
```cpp
void THCudaInit(THCState* state) {
  c10::cuda::CUDACachingAllocator::init(numDevices);
}
```

https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDACachingAllocator.cpp#L816
```cpp
  void init(int device_count) {
    int size = device_allocator.size();
    if (size < device_count) {
      device_allocator.resize(device_count);
      for (int i = size; i < device_count; i++) {
        device_allocator[i] = std::unique_ptr<DeviceCachingAllocator>(new DeviceCachingAllocator());
      }
    }
  }
```

---

I think there are no negative consequences if `init` is called multiple times. Of course, we should invoke it only once if we can.

Reply:
Fixed in #163  by using `at::globalContext().lazyInitCUDA();`.